### PR TITLE
Batch norm folding

### PR DESCRIPTION
**Pull request by jbobba**
+ ConvBiasRelu fusion in the CPU transformer

### CODE REVIEW

File: src/ngraph/pass/core_fusion.cpp [hunk 1]
```diff
@@ -73,3 +81,89 @@ void pass::CoreFusion::construct_relu()
...
+
+        auto m_bn = std::dynamic_pointer_cast<op::BatchNorm>(m.get_match_root());
+        auto m_conv = std::dynamic_pointer_cast<op::Convolution>(m_bn->get_argument(2));
+
+        if (m_conv->get_users().size() > 1)
+        {
+            return false;
+        }
+
+        if (m_conv->get_element_type() != m_bn->get_element_type())
```
[the remark this answers is not in the document]
Comment (jbobba) [reply src/ngraph/pass/core_fusion.cpp]: Mostly being paranoid here.

[the remark this answers is not in the document]
Comment (Krovatkin) [reply src/ngraph/pass/core_fusion.cpp]: let's consider this an assertion 😛

File: src/ngraph/pass/core_fusion.cpp [hunk 2]
```diff
@@ -73,3 +81,89 @@ void pass::CoreFusion::construct_relu()
...
+        auto var_eps = std::make_shared<op::Add>(
+            pattern_map[var],
+            std::make_shared<op::Broadcast>(bn_eps, pattern_map[var]->get_shape(), AxisSet{0}));
+        auto sqrt_var_eps = std::make_shared<op::Sqrt>(var_eps);
+
+        auto mean_gamma = std::make_shared<op::Multiply>(pattern_map[mean], pattern_map[gamma]);
+        auto new_biases = std::make_shared<op::Subtract>(
+            pattern_map[beta], std::make_shared<op::Divide>(mean_gamma, sqrt_var_eps));
+        auto weight_scaling = std::make_shared<op::Divide>(pattern_map[gamma], sqrt_var_eps);
+        auto new_filters = std::make_shared<op::Multiply>(
```
Comment (Krovatkin) [inline src/ngraph/pass/core_fusion.cpp]: please either rename `new_filters` to `new_weights` or vice versa so yr comment matches the logic

File: src/ngraph/runtime/cpu/pass/cpu_fusion.cpp [hunk 3]
```diff
@@ -943,3 +943,71 @@ void ngraph::runtime::cpu::pass::CPUFusion::construct_conv_relu()
...
 }
+
+void ngraph::runtime::cpu::pass::CPUFusion::construct_conv_bias_relu()
+{
+    Shape shape{2, 2, 1, 1};
+    auto data_batch = std::make_shared<pattern::op::Label>(element::f32, shape);
+    auto filters = std::make_shared<pattern::op::Label>(element::f32, shape);
+    auto bias = std::make_shared<pattern::op::Label>(element::f32, Shape{1});
+
+    auto pconv = std::make_shared<op::ConvolutionBias>(data_batch,
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_fusion.cpp]: minor gripe, can we rename `pconv` to `conv_bias` ?

File: test/cpu_fusion.cpp [hunk 4]
```diff
@@ -1223,3 +1223,62 @@ TEST(cpu_fusion, backwards_maxpool_with_indices_n4_c1_hw4_2x2_max)
...
+
+    vector<vector<float>> args{
+        {1.25f,  2.25f, 5.25f, 6.25f,  -1.25f, -1.25f, 3.25f, -4.25f, 7.25f,  8.25f,  -1.25f,
+         -1.25f, 1.25f, 2.25f, -3.25f, 2.25f,  4.25f,  4.25f, 1.25f,  2.25f,  -4.25f, 2.25f,
+         4.25f,  4.25f, 0.f,   0.f,    -1.f,   0.f,    2.f,   2.f,    0.f,    0.f,    0.f,
+         0.f,    2.f,   2.f,   1.25f,  2.25f,  5.25f,  6.25f, 1.25f,  1.25f,  3.25f,  4.25f,
+         -7.25f, 8.25f, 1.25f, -1.25f, -1.25f, 2.25f,  3.25f, 2.25f,  -4.25f, -4.25f, -1.25f,
+         -2.25f, 4.25f, 2.25f, 4.25f,  4.25f,  0.f,    0.f,   1.f,    0.f,    -2.f,   2.f,
+         0.f,    0.f,   0.f,   0.f,    -2.f,   -2.f},
+        {1.25f,
```
Comment (Krovatkin) [inline test/cpu_fusion.cpp]: I wonder if we can somehow improve formatting here? @rkimballn1

File: src/ngraph/pass/core_fusion.cpp [hunk 5]
```diff
@@ -73,3 +81,89 @@ void pass::CoreFusion::construct_relu()
     auto m = make_shared<pattern::Matcher>(max, callback);
     this->add_matcher(m);
 }
+
+void pass::CoreFusion::construct_folded_batch_norm()
+{
```
Comment (Krovatkin) [inline src/ngraph/pass/core_fusion.cpp]: this transformation seems to be geared towards MKLDNN since it's checking the rank of batchnorm. Does this really belong in CoreFusion at this point?

File: test/core_fusion.cpp [hunk 6]
```diff
@@ -26,6 +26,8 @@
 #include "ngraph/graph_util.hpp"
 #include "ngraph/log.hpp"
 #include "ngraph/ngraph.hpp"
+#include "ngraph/op/batch_norm.hpp"
+#include "ngraph/op/convolution.hpp"
```
Comment (Krovatkin) [inline test/core_fusion.cpp]: we could probably remove these since your test is in `cpu_fusion.cpp`

File: test/cpu_fusion.cpp [hunk 7]
```diff
@@ -1223,3 +1223,62 @@ TEST(cpu_fusion, backwards_maxpool_with_indices_n4_c1_hw4_2x2_max)
...
+        {-0.9384f, 0.01875f},
+        {11.0f, 1.3f},
+        {0.12f, 0.31f},
+        {0.01f, 0.11f},
+    };
+
+    auto int_results = execute(int_f, args, "INTERPRETER");
+    auto cpu_results = execute(cpu_f, args, "CPU");
+    EXPECT_TRUE(test::all_close(cpu_results.at(0), int_results.at(0)));
+}
```
Comment (Krovatkin) [inline test/cpu_fusion.cpp]: We seem to be missing tests for `conv_bias_relu` though we could do it in a separate PR.

File: src/ngraph/pass/core_fusion.cpp (hunk 5 above)
Comment (jbobba) [reply src/ngraph/pass/core_fusion.cpp]: Not really specific to MKLDNN but the replacement subgraph handles only 4-D cases, the way broadcasts are written.
